{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pip install tf-nightly\n",
    "#pip install tfp-nightly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Recommendations for single-cell analysis\n",
    "\n",
    "The DESeq2 developers and collaborating groups have published recommendations for the best use of DESeq2 for single-cell datasets, which have been described first in Van den Berge et al. (2018). Default values for DESeq2 were designed for bulk data and will not be appropriate for single-cell datasets. These settings and additional improvements have also been tested subsequently and published in Zhu, Ibrahim, and Love (2018) and Ahlmann-Eltze and Huber (2020).\n",
    "\n",
    "    Use test=\"LRT\" for significance testing when working with single-cell data, over the Wald test. This has been observed across multiple single-cell benchmarks.\n",
    "    Set the following DESeq arguments to these values: useT=TRUE, minmu=1e-6, and minReplicatesForReplace=Inf. The default setting of minmu was benchmarked on bulk RNA-seq and is not appropriate for single cell data when the expected count is often much less than 1.\n",
    "    The default size factors are not optimal for single cell count matrices, instead consider setting sizeFactors from scran::computeSumFactors.\n",
    "    One important concern for single-cell data analysis is the size of the datasets and associated processing time. To address the speed concerns, DESeq2 provides an interface to glmGamPoi, which implements faster dispersion and parameter estimation routines for single-cell data (Ahlmann-Eltze and Huber 2020). To use this feature, set fitType = \"glmGamPoi\". Alternatively, one can use glmGamPoi as a standalone package. This provides the additional option to process data on-disk if the full dataset does not fit in memory, a quasi-likelihood framework for differential testing, and the ability to form pseudobulk samples (more details how to use glmGamPoi are in its README).\n",
    "\n",
    "Optionally, one can consider using the zinbwave package to directly model the zero inflation of the counts, and take account of these in the DESeq2 model. This allows for the DESeq2 inference to apply to the part of the data which is not due to zero inflation. Not all single cell datasets exhibit zero inflation, and instead may just reflect low conditional estimated counts (conditional on cell type or cell state).There is example code for combining zinbwave and DESeq2 package functions in the zinbwave vignette. We also have an example of ZINB-WaVE + DESeq2 integration using the splatter package for simulation at the zinbwave-deseq2 GitHub repository."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "Can I use DESeq2 to analyze paired samples?\n",
    "\n",
    "Yes, you should use a multi-factor design which includes the sample information as a term in the design formula. This will account for differences between the samples while estimating the effect due to the condition. The condition of interest should go at the end of the design formula, e.g. ~ subject + condition.\n",
    "If I have multiple groups, should I run all together or split into pairs of groups?\n",
    "\n",
    "Typically, we recommend users to run samples from all groups together, and then use the contrast argument of the results function to extract comparisons of interest after fitting the model using DESeq.\n",
    "\n",
    "The model fit by DESeq estimates a single dispersion parameter for each gene, which defines how far we expect the observed count for a sample will be from the mean value from the model given its size factor and its condition group. See the section above and the DESeq2 paper for full details. Having a single dispersion parameter for each gene is usually sufficient for analyzing multi-group data, as the final dispersion value will incorporate the within-group variability across all groups.\n",
    "\n",
    "However, for some datasets, exploratory data analysis (EDA) plots could reveal that one or more groups has much higher within-group variability than the others. A simulated example of such a set of samples is shown below. This is case where, by comparing groups A and B separately – subsetting a DESeqDataSet to only samples from those two groups and then running DESeq on this subset – will be more sensitive than a model including all samples together. It should be noted that such an extreme range of within-group variability is not common, although it could arise if certain treatments produce an extreme reaction (e.g. cell death). Again, this can be easily detected from the EDA plots such as PCA described in this vignette.\n",
    "\n",
    "Here we diagram an extreme range of within-group variability with a simulated dataset. Typically, it is recommended to run DESeq across samples from all groups, for datasets with multiple groups. However, this simulated dataset shows a case where it would be preferable to compare groups A and B by creating a smaller dataset without the C samples. Group C has much higher within-group variability, which would inflate the per-gene dispersion estimate for groups A and B as well:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%reset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata\n",
    "import itertools\n",
    "import gc\n",
    "from diffexpr.py_deseq import py_DESeq2\n",
    "from rpy2.robjects import Formula\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "scaler = StandardScaler()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "q = sc.read_h5ad('../atlas/Atlas_adatas_June2021_Atlas_final_May2021.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "qobs = pd.read_csv('EsoAtlas_adatas_june2021_final_obs_May15.csv', index_col=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "qobs.diagnosis = qobs.diagnosis.astype('str')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>samplename</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_molecules</th>\n",
       "      <th>doublet_score</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>leiden</th>\n",
       "      <th>louvain</th>\n",
       "      <th>diagnosis</th>\n",
       "      <th>phase</th>\n",
       "      <th>...</th>\n",
       "      <th>hcl_celltype</th>\n",
       "      <th>hcl_score</th>\n",
       "      <th>CLid</th>\n",
       "      <th>CL_name</th>\n",
       "      <th>nobatch_leiden</th>\n",
       "      <th>nobatch_louvain</th>\n",
       "      <th>cnv_avg</th>\n",
       "      <th>has_cnv</th>\n",
       "      <th>CellLabels</th>\n",
       "      <th>Stromal</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGTTGTAAG-E12D</th>\n",
       "      <td>0</td>\n",
       "      <td>E12D</td>\n",
       "      <td>1765</td>\n",
       "      <td>3751.0</td>\n",
       "      <td>0.045362</td>\n",
       "      <td>0.024793</td>\n",
       "      <td>14</td>\n",
       "      <td>16</td>\n",
       "      <td>NS</td>\n",
       "      <td>G1</td>\n",
       "      <td>...</td>\n",
       "      <td>T.cells3.Placenta_VentoTormo.</td>\n",
       "      <td>0.378946</td>\n",
       "      <td>CL:0000084</td>\n",
       "      <td>T cell</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>NK_cells</td>\n",
       "      <td>non-stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCACATCAACCA-E12D</th>\n",
       "      <td>1</td>\n",
       "      <td>E12D</td>\n",
       "      <td>1445</td>\n",
       "      <td>2419.0</td>\n",
       "      <td>0.087379</td>\n",
       "      <td>0.045887</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NS</td>\n",
       "      <td>G1</td>\n",
       "      <td>...</td>\n",
       "      <td>T.cells3.Placenta_VentoTormo.</td>\n",
       "      <td>0.355749</td>\n",
       "      <td>CL:0000084</td>\n",
       "      <td>T cell</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>cd8_Tcells</td>\n",
       "      <td>non-stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCATCCGATCTC-E12D</th>\n",
       "      <td>2</td>\n",
       "      <td>E12D</td>\n",
       "      <td>2891</td>\n",
       "      <td>7911.0</td>\n",
       "      <td>0.046929</td>\n",
       "      <td>0.021489</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NS</td>\n",
       "      <td>G1</td>\n",
       "      <td>...</td>\n",
       "      <td>Pit.cell_TFF2.high.Adult.Stomach3.</td>\n",
       "      <td>0.502768</td>\n",
       "      <td>CL:0002180</td>\n",
       "      <td>mucous cell of stomach</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0.050446</td>\n",
       "      <td>no</td>\n",
       "      <td>gi_epithelial</td>\n",
       "      <td>non-stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAGTCAGTCTA-E12D</th>\n",
       "      <td>3</td>\n",
       "      <td>E12D</td>\n",
       "      <td>1109</td>\n",
       "      <td>1956.0</td>\n",
       "      <td>0.169611</td>\n",
       "      <td>0.032720</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NS</td>\n",
       "      <td>G1</td>\n",
       "      <td>...</td>\n",
       "      <td>T.cells3.Placenta_VentoTormo.</td>\n",
       "      <td>0.335552</td>\n",
       "      <td>CL:0000084</td>\n",
       "      <td>T cell</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>cd8_Tcells</td>\n",
       "      <td>non-stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAGTGACAACG-E12D</th>\n",
       "      <td>4</td>\n",
       "      <td>E12D</td>\n",
       "      <td>1313</td>\n",
       "      <td>2435.0</td>\n",
       "      <td>0.043875</td>\n",
       "      <td>0.021355</td>\n",
       "      <td>14</td>\n",
       "      <td>16</td>\n",
       "      <td>NS</td>\n",
       "      <td>G1</td>\n",
       "      <td>...</td>\n",
       "      <td>T.cells3.Placenta_VentoTormo.</td>\n",
       "      <td>0.345581</td>\n",
       "      <td>CL:0000084</td>\n",
       "      <td>T cell</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>NK_cells</td>\n",
       "      <td>non-stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGCAAGGCCTC-E08C</th>\n",
       "      <td>175581</td>\n",
       "      <td>E08C</td>\n",
       "      <td>2456</td>\n",
       "      <td>6679.0</td>\n",
       "      <td>0.051873</td>\n",
       "      <td>0.058841</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>T</td>\n",
       "      <td>G1</td>\n",
       "      <td>...</td>\n",
       "      <td>DC1.Placenta_VentoTormo.</td>\n",
       "      <td>0.469803</td>\n",
       "      <td>CL:0000451</td>\n",
       "      <td>dendritic cell</td>\n",
       "      <td>12</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>monocytes_macs_DCs</td>\n",
       "      <td>non-stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGCACTTTATC-E08C</th>\n",
       "      <td>175582</td>\n",
       "      <td>E08C</td>\n",
       "      <td>7132</td>\n",
       "      <td>46689.0</td>\n",
       "      <td>0.057783</td>\n",
       "      <td>0.124398</td>\n",
       "      <td>13</td>\n",
       "      <td>15</td>\n",
       "      <td>T</td>\n",
       "      <td>G2M</td>\n",
       "      <td>...</td>\n",
       "      <td>Pit.cell_TFF2.high.Adult.Stomach1.</td>\n",
       "      <td>0.515035</td>\n",
       "      <td>CL:0002180</td>\n",
       "      <td>mucous cell of stomach</td>\n",
       "      <td>37</td>\n",
       "      <td>24</td>\n",
       "      <td>0.121622</td>\n",
       "      <td>yes</td>\n",
       "      <td>gi_epithelial</td>\n",
       "      <td>non-stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTCCAGTTA-E08C</th>\n",
       "      <td>175583</td>\n",
       "      <td>E08C</td>\n",
       "      <td>908</td>\n",
       "      <td>1403.0</td>\n",
       "      <td>0.063348</td>\n",
       "      <td>0.081967</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>T.cells3.Placenta_VentoTormo.</td>\n",
       "      <td>0.267463</td>\n",
       "      <td>CL:0000084</td>\n",
       "      <td>T cell</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>cd8_Tcells</td>\n",
       "      <td>non-stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGGTGTAAATG-E08C</th>\n",
       "      <td>175584</td>\n",
       "      <td>E08C</td>\n",
       "      <td>1017</td>\n",
       "      <td>1757.0</td>\n",
       "      <td>0.043399</td>\n",
       "      <td>0.077974</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>G1</td>\n",
       "      <td>...</td>\n",
       "      <td>T.cells3.Placenta_VentoTormo.</td>\n",
       "      <td>0.368635</td>\n",
       "      <td>CL:0000084</td>\n",
       "      <td>T cell</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>cd8_Tcells</td>\n",
       "      <td>non-stromal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGTCAACGTGT-E08C</th>\n",
       "      <td>175585</td>\n",
       "      <td>E08C</td>\n",
       "      <td>1456</td>\n",
       "      <td>2991.0</td>\n",
       "      <td>0.042593</td>\n",
       "      <td>0.076897</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>G1</td>\n",
       "      <td>...</td>\n",
       "      <td>T.cells3.Placenta_VentoTormo.</td>\n",
       "      <td>0.383414</td>\n",
       "      <td>CL:0000084</td>\n",
       "      <td>T cell</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>no</td>\n",
       "      <td>cd8_Tcells</td>\n",
       "      <td>non-stromal</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175586 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Unnamed: 0 samplename  n_genes  n_molecules  \\\n",
       "Unnamed: 0.1                                                         \n",
       "AAACCCAAGTTGTAAG-E12D           0       E12D     1765       3751.0   \n",
       "AAACCCACATCAACCA-E12D           1       E12D     1445       2419.0   \n",
       "AAACCCATCCGATCTC-E12D           2       E12D     2891       7911.0   \n",
       "AAACGAAGTCAGTCTA-E12D           3       E12D     1109       1956.0   \n",
       "AAACGAAGTGACAACG-E12D           4       E12D     1313       2435.0   \n",
       "...                           ...        ...      ...          ...   \n",
       "TTTGTTGCAAGGCCTC-E08C      175581       E08C     2456       6679.0   \n",
       "TTTGTTGCACTTTATC-E08C      175582       E08C     7132      46689.0   \n",
       "TTTGTTGGTCCAGTTA-E08C      175583       E08C      908       1403.0   \n",
       "TTTGTTGGTGTAAATG-E08C      175584       E08C     1017       1757.0   \n",
       "TTTGTTGTCAACGTGT-E08C      175585       E08C     1456       2991.0   \n",
       "\n",
       "                       doublet_score  percent_mito  leiden  louvain diagnosis  \\\n",
       "Unnamed: 0.1                                                                    \n",
       "AAACCCAAGTTGTAAG-E12D       0.045362      0.024793      14       16        NS   \n",
       "AAACCCACATCAACCA-E12D       0.087379      0.045887       1        1        NS   \n",
       "AAACCCATCCGATCTC-E12D       0.046929      0.021489       0        0        NS   \n",
       "AAACGAAGTCAGTCTA-E12D       0.169611      0.032720       1        1        NS   \n",
       "AAACGAAGTGACAACG-E12D       0.043875      0.021355      14       16        NS   \n",
       "...                              ...           ...     ...      ...       ...   \n",
       "TTTGTTGCAAGGCCTC-E08C       0.051873      0.058841      10       10         T   \n",
       "TTTGTTGCACTTTATC-E08C       0.057783      0.124398      13       15         T   \n",
       "TTTGTTGGTCCAGTTA-E08C       0.063348      0.081967       1        1         T   \n",
       "TTTGTTGGTGTAAATG-E08C       0.043399      0.077974       1        1         T   \n",
       "TTTGTTGTCAACGTGT-E08C       0.042593      0.076897       1        1         T   \n",
       "\n",
       "                      phase  ...                        hcl_celltype  \\\n",
       "Unnamed: 0.1                 ...                                       \n",
       "AAACCCAAGTTGTAAG-E12D    G1  ...       T.cells3.Placenta_VentoTormo.   \n",
       "AAACCCACATCAACCA-E12D    G1  ...       T.cells3.Placenta_VentoTormo.   \n",
       "AAACCCATCCGATCTC-E12D    G1  ...  Pit.cell_TFF2.high.Adult.Stomach3.   \n",
       "AAACGAAGTCAGTCTA-E12D    G1  ...       T.cells3.Placenta_VentoTormo.   \n",
       "AAACGAAGTGACAACG-E12D    G1  ...       T.cells3.Placenta_VentoTormo.   \n",
       "...                     ...  ...                                 ...   \n",
       "TTTGTTGCAAGGCCTC-E08C    G1  ...            DC1.Placenta_VentoTormo.   \n",
       "TTTGTTGCACTTTATC-E08C   G2M  ...  Pit.cell_TFF2.high.Adult.Stomach1.   \n",
       "TTTGTTGGTCCAGTTA-E08C     S  ...       T.cells3.Placenta_VentoTormo.   \n",
       "TTTGTTGGTGTAAATG-E08C    G1  ...       T.cells3.Placenta_VentoTormo.   \n",
       "TTTGTTGTCAACGTGT-E08C    G1  ...       T.cells3.Placenta_VentoTormo.   \n",
       "\n",
       "                      hcl_score        CLid                 CL_name  \\\n",
       "Unnamed: 0.1                                                          \n",
       "AAACCCAAGTTGTAAG-E12D  0.378946  CL:0000084                  T cell   \n",
       "AAACCCACATCAACCA-E12D  0.355749  CL:0000084                  T cell   \n",
       "AAACCCATCCGATCTC-E12D  0.502768  CL:0002180  mucous cell of stomach   \n",
       "AAACGAAGTCAGTCTA-E12D  0.335552  CL:0000084                  T cell   \n",
       "AAACGAAGTGACAACG-E12D  0.345581  CL:0000084                  T cell   \n",
       "...                         ...         ...                     ...   \n",
       "TTTGTTGCAAGGCCTC-E08C  0.469803  CL:0000451          dendritic cell   \n",
       "TTTGTTGCACTTTATC-E08C  0.515035  CL:0002180  mucous cell of stomach   \n",
       "TTTGTTGGTCCAGTTA-E08C  0.267463  CL:0000084                  T cell   \n",
       "TTTGTTGGTGTAAATG-E08C  0.368635  CL:0000084                  T cell   \n",
       "TTTGTTGTCAACGTGT-E08C  0.383414  CL:0000084                  T cell   \n",
       "\n",
       "                      nobatch_leiden nobatch_louvain   cnv_avg has_cnv  \\\n",
       "Unnamed: 0.1                                                             \n",
       "AAACCCAAGTTGTAAG-E12D              1               2  0.000000      no   \n",
       "AAACCCACATCAACCA-E12D              1               2  0.000000      no   \n",
       "AAACCCATCCGATCTC-E12D              7               8  0.050446      no   \n",
       "AAACGAAGTCAGTCTA-E12D              1               2  0.000000      no   \n",
       "AAACGAAGTGACAACG-E12D              1               2  0.000000      no   \n",
       "...                              ...             ...       ...     ...   \n",
       "TTTGTTGCAAGGCCTC-E08C             12              11  0.000000      no   \n",
       "TTTGTTGCACTTTATC-E08C             37              24  0.121622     yes   \n",
       "TTTGTTGGTCCAGTTA-E08C              1               0  0.000000      no   \n",
       "TTTGTTGGTGTAAATG-E08C              1               2  0.000000      no   \n",
       "TTTGTTGTCAACGTGT-E08C              1               0  0.000000      no   \n",
       "\n",
       "                               CellLabels      Stromal  \n",
       "Unnamed: 0.1                                            \n",
       "AAACCCAAGTTGTAAG-E12D            NK_cells  non-stromal  \n",
       "AAACCCACATCAACCA-E12D          cd8_Tcells  non-stromal  \n",
       "AAACCCATCCGATCTC-E12D       gi_epithelial  non-stromal  \n",
       "AAACGAAGTCAGTCTA-E12D          cd8_Tcells  non-stromal  \n",
       "AAACGAAGTGACAACG-E12D            NK_cells  non-stromal  \n",
       "...                                   ...          ...  \n",
       "TTTGTTGCAAGGCCTC-E08C  monocytes_macs_DCs  non-stromal  \n",
       "TTTGTTGCACTTTATC-E08C       gi_epithelial  non-stromal  \n",
       "TTTGTTGGTCCAGTTA-E08C          cd8_Tcells  non-stromal  \n",
       "TTTGTTGGTGTAAATG-E08C          cd8_Tcells  non-stromal  \n",
       "TTTGTTGTCAACGTGT-E08C          cd8_Tcells  non-stromal  \n",
       "\n",
       "[175586 rows x 25 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "q.obs = qobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_design(q, qci):\n",
    "    # build design matrix\n",
    "    patient_ids = ([x[0:3] for x in qci.obs.samplename])\n",
    "    full_sample_df = pd.DataFrame({'patient':patient_ids, 'biosample':qci.obs.samplename, 'procedure':qci.obs.procedure, 'n_molecules':qci.obs.n_molecules, 'dx':qci.obs.diagnosis})\n",
    "        \n",
    "    # get average n_molecules\n",
    "    sample_df_molecules=full_sample_df.groupby('biosample').agg(avg_molecules=pd.NamedAgg(column='n_molecules', aggfunc='mean'))\n",
    "    sample_df_molecules['biosample'] = sample_df_molecules.index\n",
    "    sample_df_molecules.index = range(0,len(sample_df_molecules))\n",
    "    \n",
    "    # get the number of cells from each sample\n",
    "    cell_counts = pd.DataFrame(full_sample_df.biosample.value_counts())\n",
    "    cell_counts.columns = ['cell_counts']\n",
    "    cell_counts['biosample'] = cell_counts.index\n",
    "    \n",
    "    # merge in the cell counts and molecules\n",
    "    full_sample_df = full_sample_df.merge(cell_counts)\n",
    "    full_sample_df.drop('n_molecules', axis=1, inplace=True)\n",
    "    \n",
    "    # the list of biosamples in this cluster\n",
    "    biosample_list = list(set(full_sample_df.biosample))\n",
    "    \n",
    "    # and the order of cells as index\n",
    "    index = np.array(full_sample_df.biosample.tolist())\n",
    "    \n",
    "    # then we make the design matrix\n",
    "    sample_df = full_sample_df.drop_duplicates()\n",
    "    sample_df = sample_df.merge(sample_df_molecules)\n",
    "    sample_df.loc[:,'binned_cell_counts'] = pd.cut(sample_df.cell_counts, bins=[0,8,32,128,512,2048,100000]) #((sample_df.cell_counts - np.mean(sample_df.cell_counts)) / np.std(sample_df.cell_counts))\n",
    "    \n",
    "    return( (biosample_list, index, sample_df) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_count_matrix(biosample_list, index, qci, sample_df):\n",
    "    # sum within samples\n",
    "    res0 = pd.DataFrame()\n",
    "    for bsl in biosample_list:\n",
    "        idx = np.argwhere(index == bsl).flatten()\n",
    "        mat = qci.X[idx,:].sum(axis=0)\n",
    "        cnt_sum = mat.flatten().tolist()[0]\n",
    "        if len(res0) == 0:\n",
    "            res0 = pd.DataFrame(cnt_sum, columns=[bsl])\n",
    "        else:\n",
    "            res0 = res0.join(pd.DataFrame(cnt_sum, columns=[bsl]))\n",
    "    count_matrix = res0.loc[:, sample_df.biosample.tolist()]\n",
    "    count_matrix['id'] = qci.var.index.tolist()\n",
    "    count_matrix.index = qci.var.index\n",
    "    return(count_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "cellclusters = dict(\n",
    "    gi_epithelial=['0','3','4','6','8','13','19','20','28','31','33','34','35','36','38','39','40'], # 78049 cells, \n",
    "    squamous_epithelial=['16','18','21'],\n",
    "    fibroblasts=['7'],\n",
    "    myofibroblasts=['12'],\n",
    "    endothelial=['15','5', '30', '26'],\n",
    "    neuroendocrine=['17'],\n",
    "    parietal=['29'],\n",
    "    neutrophils=['22'],\n",
    "    monocytes_macs_DCs=['10'],  # and macs and dcs\n",
    "    B_cells=['11','23'],\n",
    "    cd4_Tcells=['2'],\n",
    "    cd8_Tcells=['1','25'],\n",
    "    NK_cells=['14'],\n",
    "    mast_cells=['9'],\n",
    "    hepatoid=['27'],\n",
    "    naive_T_cells=['24']#,\n",
    "    #stromal=['7','12','15','5', '30', '26'],\n",
    "    #gastric=['4','6']   \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gi_epithelial\n",
      "squamous_epithelial\n",
      "fibroblasts\n",
      "myofibroblasts\n",
      "endothelial\n",
      "neuroendocrine\n",
      "parietal\n",
      "neutrophils\n",
      "monocytes_macs_DCs\n",
      "B_cells\n",
      "cd4_Tcells\n",
      "cd8_Tcells\n",
      "NK_cells\n",
      "mast_cells\n",
      "hepatoid\n",
      "naive_T_cells\n"
     ]
    }
   ],
   "source": [
    "for leiden_label in cellclusters.keys():\n",
    "    print(leiden_label)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: gi_epithelial\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 612 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: squamous_epithelial\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 606 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: fibroblasts\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 507 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: myofibroblasts\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 570 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: endothelial\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 694 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: neuroendocrine\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 492 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: parietal\n",
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Error in checkFullRank(modelMatrix) : \n",
      "  the model matrix is not full rank, so the model cannot be fit as specified.\n",
      "  One or more variables or interaction terms in the design formula are linear\n",
      "  combinations of the others and must be removed.\n",
      "\n",
      "  Please read the vignette section 'Model matrix not full rank':\n",
      "\n",
      "  vignette('DESeq2')\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: In addition: \n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: There were 16 warnings (use warnings() to see them)\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error: parietal\n",
      "View of AnnData object with n_obs × n_vars = 385 × 35606\n",
      "    obs: 'Unnamed: 0', 'samplename', 'n_genes', 'n_molecules', 'doublet_score', 'percent_mito', 'leiden', 'louvain', 'diagnosis', 'phase', 'sample_diagnosis', 'patient', 'treatment', 'procedure', 'hcl_refined', 'hcl_celltype', 'hcl_score', 'CLid', 'CL_name', 'nobatch_leiden', 'nobatch_louvain', 'cnv_avg', 'has_cnv', 'CellLabels', 'Stromal'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'is_mito', 'is_ribo'\n",
      "    uns: 'leiden', 'leiden_colors', 'leiden_sizes', 'log_X', 'log_raw.X', 'louvain', 'neighbors', 'nobatch', 'nobatch_leiden_colors', 'nobatch_leiden_sizes', 'nobatch_rank_genes_groups', 'nobatch_rank_genes_groups_filtered', 'nobatch_rank_genes_groups_unfiltered', 'paga', 'pca', 'rank_genes_groups', 'rank_genes_groups_filtered', 'rank_genes_groups_unfiltered', 'umap'\n",
      "    obsm: 'X_pca', 'X_pca_original', 'X_umap', 'X_umap_nobatch'\n",
      "    obsp: 'connectivities', 'distances', 'nobatch_connectivities', 'nobatch_distances'\n",
      "\n",
      "leiden cluster: neutrophils\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 556 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: monocytes_macs_DCs\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 323 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: B_cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 1041 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: cd4_Tcells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 630 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: cd8_Tcells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 638 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: NK_cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 850 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: mast_cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 405 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: hepatoid\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 3544 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: naive_T_cells\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/2971334874.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: -- note: fitType='parametric', but the dispersion trend was not well captured by the\n",
      "   function: y = a/x + b, and a local regression fit was automatically substituted.\n",
      "   specify fitType='local' or 'mean' to avoid this message next time.\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 842 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    }
   ],
   "source": [
    "clusterlabs = list(set(q.obs.leiden))\n",
    "clusterlabs.sort()\n",
    "res_df = pd.DataFrame()\n",
    "for leiden_label in cellclusters.keys(): #clusterlabs:\n",
    "    # subset the anndata to this cluster\n",
    "    print('leiden cluster: ' + leiden_label)\n",
    "    ### subset data to this cluster\n",
    "    qci = q[q.obs.CellLabels == leiden_label]\n",
    "    qci = qci[qci.obs.diagnosis.isin(['NE', 'NS', 'M', 'D', 'T'])]\n",
    "    ###\n",
    "    (biosample_list, index, sample_df) = build_design(q, qci)\n",
    "    try:\n",
    "        # building the pseudobulk count matrix\n",
    "        count_matrix = build_count_matrix(biosample_list, index, qci, sample_df)\n",
    "        #sample_df.binned_cell_counts = [str(x) for x in sample_df.binned_cell_counts]\n",
    "        sample_df.index = sample_df.biosample\n",
    "        x =  scaler.fit_transform(sample_df[ ['cell_counts', 'avg_molecules'] ])\n",
    "        sample_df.cell_counts = x[:,0]\n",
    "        sample_df.avg_molecules = x[:,1]\n",
    "        sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n",
    "        sample_df = sample_df[['patient','cell_counts','avg_molecules','dx']]\n",
    "        #fit a deseq2 model\n",
    "        #https://bioconductor.riken.jp/packages/3.6/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#group-specific-condition-effects-individuals-nested-within-groups\n",
    "        print(' .... running deseq2...')\n",
    "        ###\n",
    "        dds = py_DESeq2(count_matrix = count_matrix,\n",
    "                       design_matrix = sample_df,\n",
    "                       design_formula = '~ patient + cell_counts + avg_molecules + dx', #'~ patient + cell_counts + dx',\n",
    "                       gene_column = 'id') # <- telling DESeq2 this should be the gene ID column\n",
    "        params = dict(test='LRT', reduced=Formula('~ patient + cell_counts + avg_molecules'), useT=True, minmu=1e-6, minReplicatesForReplace=np.Inf) # \n",
    "        dds.run_deseq(**params) \n",
    "        ### then pulling out log2FC\n",
    "        dxs = list(set(sample_df.dx))\n",
    "        for (dx1, dx2) in list(itertools.product(dxs, dxs)):\n",
    "            if dx1 != dx2:\n",
    "                dds.get_deseq_result(contrast = ['dx',dx1, dx2])\n",
    "                de_df = dds.deseq_result \n",
    "                # add the additional items\n",
    "                de_df['celltype'] = leiden_label\n",
    "                #sig_res = res_df[(res_df.padj < 0.05) & (abs(res_df.log2FoldChange) > 1) & (res_df.baseMean > 2)]\n",
    "                #sig_res[ '_'.join(['dx','D','T']) ] = sig_res.log2FoldChange\n",
    "                de_df.to_csv('deseq2_out/deseq2_batch_'+str(leiden_label)+'_'+dx1+'_'+dx2+'.csv')\n",
    "    except:\n",
    "        print('error: ' + leiden_label)\n",
    "        print(qci)\n",
    "        print('')\n",
    "    del qci\n",
    "    gc.collect()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "leiden cluster: stromal\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_599/3824270855.py:20: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only\n",
      "  sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " .... running deseq2...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: converting counts to integer mode\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating size factors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: estimating dispersions\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: mean-dispersion relationship\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: final dispersion estimates\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: fitting model and testing\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: 623 rows did not converge in beta, labelled in mcols(object)$fullBetaConv. Use larger maxit argument with nbinomLRT\n",
      "\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'T', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'D', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NS']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'M', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NS', 'NE']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'T']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'D']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'M']\n",
      "INFO:DESeq2:Using contrast: ['dx', 'NE', 'NS']\n"
     ]
    }
   ],
   "source": [
    "# And running on stromal cells\n",
    "\n",
    "clusterlabs = list(set(q.obs.leiden))\n",
    "clusterlabs.sort()\n",
    "res_df = pd.DataFrame()\n",
    "for leiden_label in ['stromal']: #clusterlabs:\n",
    "    # subset the anndata to this cluster\n",
    "    print('leiden cluster: ' + leiden_label)\n",
    "    ### subset data to this cluster\n",
    "    qci = q[q.obs.Stromal == leiden_label]\n",
    "    qci = qci[qci.obs.diagnosis.isin(['NE', 'NS', 'M', 'D', 'T'])]\n",
    "    ###\n",
    "    (biosample_list, index, sample_df) = build_design(q, qci)\n",
    "    try:\n",
    "        # building the pseudobulk count matrix\n",
    "        count_matrix = build_count_matrix(biosample_list, index, qci, sample_df)\n",
    "        #sample_df.binned_cell_counts = [str(x) for x in sample_df.binned_cell_counts]\n",
    "        sample_df.index = sample_df.biosample\n",
    "        x =  scaler.fit_transform(sample_df[ ['cell_counts', 'avg_molecules'] ])\n",
    "        sample_df.cell_counts = x[:,0]\n",
    "        sample_df.avg_molecules = x[:,1]\n",
    "        sample_df.drop(['binned_cell_counts', 'procedure', 'biosample'], 1, inplace=True)\n",
    "        sample_df = sample_df[['patient','cell_counts','avg_molecules','dx']]\n",
    "        #fit a deseq2 model\n",
    "        #https://bioconductor.riken.jp/packages/3.6/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#group-specific-condition-effects-individuals-nested-within-groups\n",
    "        print(' .... running deseq2...')\n",
    "        ###\n",
    "        dds = py_DESeq2(count_matrix = count_matrix,\n",
    "                       design_matrix = sample_df,\n",
    "                       design_formula = '~ patient + cell_counts + avg_molecules + dx', #'~ patient + cell_counts + dx',\n",
    "                       gene_column = 'id') # <- telling DESeq2 this should be the gene ID column\n",
    "        params = dict(test='LRT', reduced=Formula('~ patient + cell_counts + avg_molecules'), useT=True, minmu=1e-6, minReplicatesForReplace=np.Inf) # \n",
    "        dds.run_deseq(**params) \n",
    "        ### then pulling out log2FC\n",
    "        dxs = list(set(sample_df.dx))\n",
    "        for (dx1, dx2) in list(itertools.product(dxs, dxs)):\n",
    "            if dx1 != dx2:\n",
    "                dds.get_deseq_result(contrast = ['dx',dx1, dx2])\n",
    "                de_df = dds.deseq_result \n",
    "                # add the additional items\n",
    "                de_df['celltype'] = leiden_label\n",
    "                #sig_res = res_df[(res_df.padj < 0.05) & (abs(res_df.log2FoldChange) > 1) & (res_df.baseMean > 2)]\n",
    "                #sig_res[ '_'.join(['dx','D','T']) ] = sig_res.log2FoldChange\n",
    "                de_df.to_csv('deseq2_out/deseq2_batch_'+str(leiden_label)+'_'+dx1+'_'+dx2+'.csv')\n",
    "    except:\n",
    "        print('error: ' + leiden_label)\n",
    "        print(qci)\n",
    "        print('')\n",
    "    del qci\n",
    "    gc.collect()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "name": "r-cpu.4-0.m75",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/r-cpu.4-0:m75"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
